components/eamxx/src/diagnostics/field_at_height.cpp

```diff
@@ -0,0 +1,164 @@
+#include "diagnostics/field_at_height.hpp"
+
+#include "ekat/std_meta/ekat_std_utils.hpp"
+#include "ekat/util/ekat_units.hpp"
+
+namespace
+{
+// Find first position in array pointed by [beg,end) that is below z
+// If all z's in array are >=z, return end
+template<typename T>
+KOKKOS_INLINE_FUNCTION
+const T* find_first_smaller_z (const T* beg, const T* end, const T& z)
+{
+  int count = end - beg;
+  if (count==1) {
+    return z < *beg ? beg : end;
+  } else {
+    auto mid = beg + count/2;
+    return z>= *mid ? find_first_smaller_z (mid,end,z)
```

Comment:
Recursion on the GPU is generally troublesome because the compiler can't know in advance how many registers to allocate.

Reply:
I can turn this into a while loop, like in ekat's upper_bound (which, in case someone wondered, I could not use (without hacking it) b/c z is decreasing rather than increasing).